### PR TITLE
Add the basic fit engine

### PR DESCRIPTION
**Pull request by AnthonyLim23**
This PR changes the API and the way fitting is handled. The new method has a generic fitting engine that can then wrap specific minimizers (e.g. scipy). 

The new API has refactored out repeated code from the 3 existing workflows and reduced the user API to simple function calls. The user will need to inherit the workflow object and override some of the methods.

To test, just make sure that the tests pass.

### CODE REVIEW

**Review by SilkeSchomann — CHANGES_REQUESTED** [review]
Some of my change requests are proper nit picking, just some typos. There are a few methods that could be made static and also a few code duplications that could probably moved to a helper class and then just shared.

File: src/quasielasticbayes/v2_python/fit_engines/scipy_fit_engine.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+        :param upper: the upper bound for the function parameters
+        """
+        # validate
+        if len(guess) != len(upper) or len(upper) != len(lower):
+            raise ValueError(f"The guess {guess}, lower {lower} and "
+                             f"upper {upper} bounds must "
+                             "be the same length")
+        self._guess = guess
+        self._lower = lower
+        self._upper = upper
```
[the remark this answers is not in the document]
Comment (AnthonyLim23) [reply src/quasielasticbayes/v2_python/fit_engines/scipy_fit_engine.py:51]: It was originally, then I realised that I wanted the same code to be called elsewhere. So replaced the init version with a function call to here https://github.com/mantidproject/quasielasticbayes/pull/50/files#diff-9b676f5ceaca9d2686ac42a28f5dbd45e565c1cb7caa67ea11fcd39a57315155R31

File: src/quasielasticbayes/v2_python/workflows/muon_exp_decay_main.py [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+    engine = workflow.fit_engine
+    fits = []
+    errors_fit = []
+    x_data = []
+    for j in range(max_features):
+        x_data, y, e, df, de = engine.get_fit_values(j)
+        fits.append(y)
+        errors_fit.append(e)
+
+    return results, results_errors, x_data, fits, errors_fit
```
[the remark this answers is not in the document]
Comment (AnthonyLim23) [reply src/quasielasticbayes/v2_python/workflows/muon_exp_decay_main.py:93]: These are more examples for users. I would expect users to write their own version of this code (different fit engines, possibly even different fitting functions. 

I have also kept these to replicate the original methods so we can keep the same tests.

File: test/fit_engines/fitengine_test.py [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+    def get_spline_fits(self):
+        expected_y = [.095, 0.188, 0.281, 0.374, 0.467, 0.560,
+                      0.653, 0.746, 0.839, 0.932]
+        expected_e = [0.022, 0.019, 0.016, 0.013, 0.012, 0.012,
+                      0.013, 0.015, 0.017, 0.020]
+        expected_d = [-0.168, 0.046, -0.095, -0.185, -0.044, -0.160,
+                      0.017, -0.130, -0.001, -0.025]
+        expected_de = [0.055, 0.053, 0.052, 0.052, 0.051, 0.051,
+                       0.052, 0.052, 0.053, 0.054]
+        return expected_y, expected_e, expected_d, expected_de
```
[the remark this answers is not in the document]
Comment (AnthonyLim23) [reply test/fit_engines/fitengine_test.py:72]: that is a fluke to be honest. It just happens that this is using scipy too. Yes I can refactor it

File: src/quasielasticbayes/v2_python/workflows/muon_exp_decay_main.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (AnthonyLim23) [reply src/quasielasticbayes/v2_python/workflows/muon_exp_decay_main.py:93]: Also the workflow called is different (line 73) and the max_features are different (muon=4, qldata=3, qse=1) , but that is the only difference.

File: src/quasielasticbayes/v2_python/fit_engines/scipy_fit_engine.py [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+    A scipy curve fit fit engine.
+    This will use scipy's curve fit to
+    fit data.
+    """
+
+    def __init__(self, x_data: ndarray, y_data: ndarray, e_data: ndarray,
+                 lower: ndarray, upper: ndarray, guess: ndarray,
+                 max_iterations: int = 220000):
+        """
+        Creates the scipy curve fit historty class
```
[the remark this answers is not in the document]
Comment (AnthonyLim23) [reply src/quasielasticbayes/v2_python/fit_engines/scipy_fit_engine.py:18]:
```suggestion
        Creates the scipy curve fit engine class
```

**Pull request merged**